File: day-2/main.go
package main

import (
	"bufio"
	"math"
	"os"
	"strconv"
	"strings"
)

func ConvertToInteger(s string) int {
	val, err := strconv.Atoi(s)

	if err != nil {
		panic(err)
	}

	return val
}

func main() {
	file, err := os.Open("day-2/input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	count := 0
	for scanner.Scan() {
		var fields = strings.Fields(scanner.Text())

		var seriesIncreasing bool
		seriesInitialized := false
		for i := 1; i < len(fields); i++ {

			first, second := ConvertToInteger(fields[i-1]), ConvertToInteger(fields[i])

			difference := second - first

			if math.Abs(float64(difference)) >= 1 && math.Abs(float64(difference)) <= 3 {
				pairIncreasing := second > first

				if seriesInitialized {
					if pairIncreasing != seriesIncreasing {
						break
					}

					if i == len(fields)-1 {
						count++
					}
				} else {
					seriesInitialized = true
					seriesIncreasing = pairIncreasing
				}
			} else {
				break
			}
		}
	}

	print("Safe reports: ", count)

	if err := scanner.Err(); err != nil {
		panic(err)
	}

}

File: day-2/main.py
file = open("day-2/input.txt")


def isSafeReport(numbers, recursive):
  seriesPattern = None

  for i in range(len(numbers) - 1):

    first, second = numbers[i], numbers[i + 1]

    pairPattern = second > first

    difference = abs(second - first)

    if difference >= 1 and difference <= 3:
      if seriesPattern is None:
        seriesPattern = pairPattern
      elif seriesPattern != pairPattern:
        if recursive:
          return False
        firstAns = isSafeReport(numbers[:i] + numbers[i + 1:],
                                True) or isSafeReport(
                                    numbers[:i + 1] + numbers[i + 2:], True)
        if i - 1 == 0:
          secondAns = isSafeReport(numbers[1:], True)
          return firstAns or secondAns
        return firstAns
      elif i == len(numbers) - 2:
        return True
    else:
      if recursive:
        return False

      return isSafeReport(numbers[:i] + numbers[i + 1:], True) or isSafeReport(
          numbers[:i + 1] + numbers[i + 2:], True)


count = 0
for line in file.readlines():
  numbers = line.split()
  numbers = list(map(int, numbers))

  if isSafeReport(numbers, False):
    count += 1

print(count)
